internal/blockchain/fabric/fabric_provider.go

```diff
@@ -0,0 +1,297 @@
+	if !member.External {
+		return fmt.Sprintf("http://fabconnect_%s:3000", member.ID)
+	} else {
+		return fmt.Sprintf("http://127.0.0.1:%v", member.ExposedEthconnectPort)
+	}
+}
+
+func (p *FabricProvider) writeConfigtxYaml() error {
+	filePath := path.Join(constants.StacksDir, p.Stack.Name, "blockchain", "configtx.yaml")
+	return ioutil.WriteFile(filePath, []byte(configtxYaml), 0755)
+}
+
+func (p *FabricProvider) createChannel() error {
+	p.Log.Info("creating channel")
+	stackDir := path.Join(constants.StacksDir, p.Stack.Name)
+	volumeName := fmt.Sprintf("%s_firefly_fabric", p.Stack.Name)
+	return docker.RunDockerCommand(stackDir, p.Verbose, p.Verbose, "run", "--rm", fmt.Sprintf("--network=%s_default", p.Stack.Name), "-v", fmt.Sprintf("%s:/etc/firefly", volumeName), "hyperledger/fabric-tools:2.3", "osnadmin", "channel", "join", "--channelID", "firefly", "--config-block", "/etc/firefly/firefly.block", "-o", "fabric_orderer:7053", "--ca-file", "/etc/firefly/organizations/ordererOrganizations/example.com/users/Admin@example.com/tls/ca.crt", "--client-cert", "/etc/firefly/organizations/ordererOrganizations/example.com/users/Admin@example.com/tls/client.crt", "--client-key", "/etc/firefly/organizations/ordererOrganizations/example.com/users/Admin@example.com/tls/client.key")
+}
+
+func (p *FabricProvider) joinChannel() error {
+	p.Log.Info("joining channel")
+	stackDir := path.Join(constants.StacksDir, p.Stack.Name)
+	volumeName := fmt.Sprintf("%s_firefly_fabric", p.Stack.Name)
+	return docker.RunDockerCommand(stackDir, p.Verbose, p.Verbose, "run", "--rm", fmt.Sprintf("--network=%s_default", p.Stack.Name), "-v", fmt.Sprintf("%s:/etc/firefly", volumeName), "-e", "CORE_PEER_ADDRESS=fabric_peer:7051", "-e", "CORE_PEER_TLS_ENABLED=true", "-e", "CORE_PEER_TLS_ROOTCERT_FILE=/etc/firefly/organizations/peerOrganizations/org1.example.com/peers/fabric_peer.org1.example.com/tls/ca.crt", "-e", "CORE_PEER_LOCALMSPID=Org1MSP", "-e", "CORE_PEER_MSPCONFIGPATH=/etc/firefly/organizations/peerOrganizations/org1.example.com/users/Admin@org1.example.com/msp", "hyperledger/fabric-tools:2.3", "peer", "channel", "join", "-b", "/etc/firefly/firefly.block")
+}
+
+func (p *FabricProvider) extractChaincode() error {
+	stackDir := path.Join(constants.StacksDir, p.Stack.Name)
+	contractsDir := path.Join(stackDir, "contracts")
+
+	if err := os.MkdirAll(contractsDir, 0755); err != nil {
+		return err
+	}
+
+	var containerName string
+	for _, member := range p.Stack.Members {
+		if !member.External {
+			containerName = fmt.Sprintf("%s_firefly_core_%s", p.Stack.Name, member.ID)
+			break
+		}
+	}
+	if containerName == "" {
+		return errors.New("unable to extract contracts from container - no valid firefly core containers found in stack")
+	}
+	p.Log.Info("extracting smart contracts")
+	if err := docker.CopyFromContainer(containerName, "/firefly/contracts/firefly_fabric.tar.gz", path.Join(contractsDir, "firefly_fabric.tar.gz"), p.Verbose); err != nil {
```

Comment:
Is this tarball coming from the local CLI directory or from a container? I saw a tarball checked in above.